src/server/page_server.cpp

```diff
@@ -244,6 +258,75 @@ page_server::connection_handler::prior_sender_sink_hook (std::string &&message)
   m_conn->push (page_to_tran_request::SEND_TO_PTS_LOG_PRIOR_LIST, std::move (message));
 }
 
+page_server::disconnect_handler::disconnect_handler ()
+  : m_terminate { false }
+{
+  m_thread = std::thread (&page_server::disconnect_handler::disconnect_loop, std::ref (*this));
+}
+
+page_server::disconnect_handler::~disconnect_handler ()
+{
+  if (m_terminate.load () && m_thread.joinable ())
+    {
+      m_queue_cv.notify_one ();
+      m_thread.join ();
+    }
+  else
+    {
+      assert (m_terminate.load ());
+    }
+
+  assert (m_disconnect_queue.empty ());
+}
+
+void
+page_server::disconnect_handler::disconnect (connection_handler_uptr_t &&handler)
+{
+  if (!m_terminate.load ())
+    {
+      std::lock_guard<std::mutex> lockg { m_queue_mtx };
+      m_disconnect_queue.emplace (std::move (handler));
+      m_queue_cv.notify_one ();
+    }
+  else
+    {
+      // cannot ask for disconnect after termination
+      assert (false);
+    }
+}
+
+void
+page_server::disconnect_handler::terminate ()
+{
+  m_terminate.store (true);
```

Comment:
Do we really need this function? How about moving "m_terminate.store (true)" into destructor and get rid og terminate()?